[🐛 Bug]: Selenium 4.32.0 ruby bindings fail to start firefox in parallel testing

### Description

Selenium 4.32.0 attempts to address issues with Firefox 136 and later failing when more that one instance is started (see https://github.com/mozilla/geckodriver/issues/2218 for details) by allocating a unique bidi port for each instance in #15458.

That works fine if all the instances are started in a single thread but fails if two threads are starting instances simultaneously as `PortProber` works by allocating and then immediately closing a port leaving another thread free to do the same thing in the gap before firefox starts and binds it.

### Reproducible Code

```shell
Not available
```

### Debugging Logs


I added `puts command.inspect` to the `build_process` method in `ServiceManager` in order to see what processes were being started and then ran my test suite single threaded:

```logs
% PARALLEL_WORKERS=1 bundle exec rails test:system

Run options: --seed 22623

# Running:

["/usr/bin/geckodriver", "--port=4444", "--websocket-port", "9223"]
.["/usr/bin/geckodriver", "--port=4445", "--websocket-port", "9224"]
............................................................["/usr/bin/geckodriver", "--port=4446", "--websocket-port", "9225"]
...................................................................................................................

Finished in 110.831997s, 1.5880 runs/s, 7.3084 assertions/s.
176 runs, 810 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for Minitest to /home/tom/src/osm/rails/coverage.
Line Coverage: 46.57% (3587 / 7702)
Lcov style coverage report generated for Minitest to coverage/lcov.info
```

and then I repeated the same with multiple threads:

```logs
% bundle exec rails test:system
Running 176 tests in parallel using 24 processes
Run options: --seed 57887

# Running:

["/usr/bin/geckodriver", "--port=4444", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4445", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4446", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4447", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4448", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4449", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4450", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4451", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4452", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4453", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4454", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4455", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4456", "--websocket-port", "9225"]
.["/usr/bin/geckodriver", "--port=4457", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4458", "--websocket-port", "9225"]
.["/usr/bin/geckodriver", "--port=4459", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4460", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4461", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4462", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4463", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4464", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4465", "--websocket-port", "9225"]
.["/usr/bin/geckodriver", "--port=4466", "--websocket-port", "9225"]
["/usr/bin/geckodriver", "--port=4467", "--websocket-port", "9225"]
.....E

Error:
IndexTest#test_should_not_add_overlay_when_not_closing_right_menu_popup:
Selenium::WebDriver::Error::NoSuchWindowError: Browsing context has been discarded
    test/system/index_test.rb:26:in 'block in <class:IndexTest>'
```

As you can see in the second case multiple instances were started using the same bidi port and tests then started failing en mass as most of the firefox instances had failed to start.

---
ℹ️ **Last known working version:** `` 4.31.0 with firefox < 136 or a monkey patch to add `--websocket-port=0` ``

I blew away `~/.cache/bazel` which seems to have fixed the ccache issue but that just moves the problem:

```
In file included from ./enc/euc_jp.c:503:
enc/jis/props.kwd:146:1: error: conflicting types for ‘onig_jis_property’; have ‘const struct enc_property *(const char *, size_t)’ {aka ‘const struct enc_property *(const char *, long unsigned int)’}
enc/jis/props.kwd:40:35: note: previous declaration of ‘onig_jis_property’ with type ‘const struct enc_property *(void)’
   40 | static const struct enc_property *onig_jis_property(/*!ANSI{*/const char *str, unsigned int len/*}!ANSI*/);
      |                                   ^~~~~~~~~~~~~~~~~
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
make[1]: *** [enc.mk:645: enc/euc_jp.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [uncommon.mk:888: enc] Error 2
make: *** Waiting for unfinished jobs....
libffi_version: 3.4.6

BUILD FAILED (Fedora Linux 42 on x86_64 using ruby-build 20250215)
```

which I fear is because it's trying to compile ruby 3.1 (why a three old version? why are we building ruby from source when I have a perfectly good one?) with gcc 15 and is running into https://gcc.gnu.org/gcc-15/porting_to.html#c23-fn-decls-without-parameters.

--- a/rb/lib/selenium/webdriver/firefox/service.rb
+++ b/rb/lib/selenium/webdriver/firefox/service.rb
@@ -28,9 +28,9 @@ module Selenium
 
         def initialize(path: nil, port: nil, log: nil, args: nil)
           args ||= []
-          unless args.any? { |arg| arg.include?('--connect-existing') }
+          unless args.any? { |arg| arg.include?('--connect-existing') || arg.include?('--websocket-port') }
             args << '--websocket-port'
-            args << WebDriver::PortProber.above(9222).to_s
+            args << '0'
           end
           super
         end

--- a/rb/spec/integration/selenium/webdriver/firefox/service_spec.rb
+++ b/rb/spec/integration/selenium/webdriver/firefox/service_spec.rb
@@ -39,21 +39,6 @@ module Selenium
         it 'can be started outside driver' do
           expect(service_manager.uri).to be_a(URI)
         end
-
-        context 'with BiDi enabled', exclusive: {bidi: true, reason: 'only executed when bidi is enabled'} do
-          it 'ensures two service instances use different websocket port' do
-            service1 = described_class.new
-            service2 = described_class.new
-
-            ws_index1 = service1.args.index('--websocket-port')
-            ws_index2 = service2.args.index('--websocket-port')
-
-            port1 = service1.args[ws_index1 + 1].to_i
-            port2 = service2.args[ws_index2 + 1].to_i
-
-            expect(port1).not_to eq(port2)
-          end
-        end
       end
     end # Firefox
   end # WebDriver

--- a/rb/spec/unit/selenium/webdriver/firefox/service_spec.rb
+++ b/rb/spec/unit/selenium/webdriver/firefox/service_spec.rb
@@ -77,11 +77,11 @@ module Selenium
             expect(service.extra_args).to include(*%w[--foo --bar])
           end
 
-          it 'there is a random port for websocket' do
+          it 'there is a zero port for websocket' do
             service = described_class.new
             ws_index = service.extra_args.index('--websocket-port')
             port = service.extra_args[ws_index + 1].to_i
-            expect(port).to be_positive
+            expect(port).to be_zero
           end
 
           context 'with connect existing' do
@@ -89,6 +89,14 @@ module Selenium
               service = described_class.new(args: ['--connect-existing'])
               expect(service.extra_args).not_to include('--websocket-port')
               expect(service.extra_args).to eq(['--connect-existing'])
+            end
+          end
+
+          context 'with websocket port' do
+            it 'does not add websocket-port' do
+              service = described_class.new(args: ['--websocket-port=1234'])
+              expect(service.extra_args).not_to include('--websocket-port=0')
+              expect(service.extra_args).to eq(['--websocket-port=1234'])
             end
           end
         end

